SS-1: First pass at the ADRECS brick, Need help with DVC pipeline

Sulstice opened the pull request:
Hi,

In this PR, I've added the module `ADRECS` into the bricks layout. The module is simple with two functions `download_data` and `process_data`. It is turned into a click command pipeline. 

The file formats the data is ingested in is `gzip` and `xlsx` in byte format. I outputted into the parquet files. 

@zmughal 

I added the `dvc.yaml` file and you can run the code with `dvc repro` however I can't figure out the correct file pathing with the yaml file and wondering if you can help me. Why is there `raw` and `list` in some and what does that mean? 

```python

    deps:
      - ./stages/01_adrecs.py
    outs:
      - brick
```

I can probably figure it out soon but worth putting up the PR now.

Sulstice committed:
added click pipeline commands for dvc

Sulstice committed:
added the gitignore for python dist and cahcing, requirments, brick directory path,

Sulstice committed:
added the dvc brick

Sulstice committed:
added the dvc start of the code

zmughal commented on .gitignore:
```diff
@@ -2,3 +2,137 @@ logs
 /download
 /list
 /brick
+.idea
```
Many of these should go in a global `.gitignore`: <https://sebastiandedeyne.com/setting-up-a-global-gitignore-file/> rather than added to the project. That keeps the files clean of things that aren't part of the function of the project (YAGNI).

Sulstice replied:
Got it, my own personal `.gitignore` file as compared to a project based one. Makes sense.

zmughal commented on stages/01_adrecs.py:
```diff
@@ -1,21 +1,45 @@
 
+@click.command()
+@click.option('--download_data', default=False, help='Run the downloading of the ADRECS data')
+@click.option('--process_data', default=False, help='Process the Data into parquet files')
```
You should probably use `is_flag=True` on both of these as that is the Unix CLI convention for booleans. That would mean making the command line just: `stages/01_adrecs.py --download_data` (remove the `True`).

Sulstice replied:
Maybe we can chat about making a command line pipeline for the bricks if we need too. For now, since most bricks are not built in that way I don't think I should start with a different way. 

Resolved in this commit to split the files.

57a5a6bf583488af0ef0743a5a01181957ba812f

zmughal commented on stages/01_adrecs.py:
```diff
@@ -1,21 +1,45 @@
+  elif process_data:
+    adrecs.process_data()
+  else:
+    pass
```
This last part should be an error. `raise` a [`UsageError` here](https://pocco-click.readthedocs.io/en/latest/exceptions.html#which-exceptions-exist).

zmughal commented on stages/01_adrecs.py:
```diff
@@ -1,21 +1,45 @@
+        '''
+
+        os.makedirs('cache', exist_ok=True)
+        os.makedirs('brick', exist_ok=True)
```
Only create the `cache` directory here. You will need to create the `brick` directory when running `process_data`.

zmughal commented on stages/01_adrecs.py:
```diff
@@ -27,12 +51,18 @@ def retrieve_data(self):
+        '''
+
+        _, _, files = next(os.walk('cache'))
+        files = [ os.path.join('brick', file) for file in files ]
```
`files` variable contains the files under `cache`, so to access them, the `join()` would be `join('cache', file)`.

zmughal replied:
There's a cleaner approach if you use `pathlib.Path`. Given a file `f` under the `cache/` directory, you can do:

```ipython
In [1]: from pathlib import Path

In [2]: cache_dir, brick_dir = Path('cache'), Path('brick')

In [3]: f = Path('cache/path/to/one.csv')

In [4]: brick_dir / f.relative_to(cache_dir).with_suffix('.parquet')
Out[4]: PosixPath('brick/path/to/one.parquet')
```

Note how this changes the parent path prefix and the file suffix.

zmughal commented on dvc.yaml:
```diff
@@ -3,33 +3,19 @@
 stages:
   download:
-    cmd: stages/01_download.sh
+    cmd: python stages/01_adrecs.py --download_data True
```
I would use `python3` here. See <https://peps.python.org/pep-0394/>.

zmughal commented on dvc.yaml:
```diff
@@ -3,33 +3,19 @@
     deps:
-      - stages/01_download.sh
-    outs:
-      - download
```
You're using `cache` as the directory name. I would use `download`. With `dvc` all steps are cached. You will need to indicate the directory under the `outs`.

zmughal commented on dvc.yaml:
```diff
@@ -3,33 +3,19 @@
-      - stages/csv2parquet.sh
-      - raw
-      - list
+      - ./stages/01_adrecs.py
```
This also has a dependency on the output directory of the `download` stage.

zmughal reviewed (changes requested):
While it may be fine for this brick because there are so few small files, I wouldn't use a single script as you lose the advantage of using `dvc` in that it will keep the downloads cached while iterating on the pipeline. If you use a single script to do both the download and processing of the data, then each change to the script will redownload the data when using `dvc repro` even if all that is being changed is how the data is processed.

So to make future refactoring easier, I would recommend creating a script-per-stage. This also removes the need for using the check for `__main__` and using `click` because it is just a straight script.

Also, I don't think it is necessary here, but you might be able to use `wget` for the `status` stage in #4 and the `download` stage. You can take a look at other bricks for how they use `wget` for examples.

zmughal commented:
Just to show the `wget` approach:

```shell
wget -r \
       --accept-regex '.*/ADReCS/download/.*' \
       -P download \
       --no-host-directories --cut-dirs=2 \
       'http://bioinf.xmu.edu.cn/ADReCS/download.jsp'
```

zmughal commented on stages/01_adrecs.py:
```diff
@@ -27,12 +51,18 @@ def retrieve_data(self):
+        _, _, files = next(os.walk('cache'))
+        files = [ os.path.join('brick', file) for file in files ]
 
         for file in files:
```
You should be able to do:
```python
df = pd.read_csv('download/v3.3/Drug_ADR_v3.3.txt.gz', sep='\t')
```
No need for decompression and manually extracting the headers.

Sulstice replied:
There was some encode if you mean why I did the decode('utf-8') and popped the headers? or iterating through the directory?

zmughal replied:
`read_csv()` can also take an `encoding` parameter which defaults to `utf-8`: <https://pandas.pydata.org/pandas-docs/version/2.2/reference/api/pandas.read_csv.html>.

Sulstice committed:
Changed the code to be functional and not module split sections

Sulstice committed:
Added the DVC file and the stages for commit

Sulstice committed:
Changed the directory name for downloads

Sulstice committed:
Changed the gitignore file

zmughal commented on stages/02_process_data.py:
```diff
@@ -0,0 +1,32 @@
+import pandas as pd
+
+_, _, files = next(os.walk('downloads'))
+files = [ os.path.join('downloads', file) for file in files ]
```
Try the following:

```python
files = filter( lambda item: item.is_file(), pathlib.Path('download').rglob('*') )
```

zmughal commented on stages/02_process_data.py:
```diff
@@ -0,0 +1,32 @@
+              lines = file_in.read().decode("utf-8").split('\n')
+              data = [ line.split('\t') for line in lines ]
+              headers = data.pop(0)
+              df = pd.DataFrame(data, columns=headers)
```
```python
df = pd.read_csv(file, sep='\t', encoding='utf-8')
```

Sulstice replied:
I did try that:

```python
ADRECS File Conversion Failed: Error converting column "BADD_A00003" to bytes using encoding UTF8. Original error: bad argument type for built-in operation
```

I would get this.

Sulstice replied:
I might pass on that one.

zmughal replied:
On which file?

Sulstice replied:
```
http://bioinf.xmu.edu.cn/ADReCS/download/v3.3/Drug_information_v3.3.xlsx
```

zmughal replied:
That's odd, it's working here:

```python
In [15]: import pandas, openpyxl; ( pandas.__version__, openpyxl.__version__ )
Out[15]: ('2.2.2', '3.1.2')

In [16]: pd.read_excel('download/v3.3/Drug_information_v3.3.xlsx', engine='openpyxl').count()
Out[16]:
DRUG_ID          2601
DRUG_NAME        2601
DRUG_SYNONYMS    2601
DrugBank_ID      2601
PubChem_ID       2601
MESH_ID          2601
KEGG_ID          2601
TTD_ID           2601
dtype: int64
```

Sulstice replied:
Sorry wrong file: http://bioinf.xmu.edu.cn/ADReCS/download/v3.3/Drug_ADR_v3.3.txt.gz

zmughal replied:
It was `ADReCS_ADR_Severity_Grade_v3.3.txt.gz` because of the comment below.

zmughal commented on stages/02_process_data.py:
```diff
@@ -0,0 +1,32 @@
+
+for file in files:
+
+  out_file = os.path.join('brick', file.split('/')[1].split('.')[0] + '.parquet')
```
```python
out_file = brick_dir / f.relative_to(download_dir).with_suffix('.parquet')
```

This makes it more clear what is happening. Also handles paths with a depth greater than one under `download`.

zmughal replied:
Hmm, you might have to do something more along the lines of:

```python
brick_dir / f.relative_to(cache_dir).with_name( re.sub( r'\.(txt\.gz|xlsx)$', '.parquet', f.name ) )
```
because `.with_suffix()` can only remove the final suffix component.

Sulstice replied:
I think this is a little too complex. It's pretty simple with the Posix Path object I think and easy to scale.

zmughal replied:
If the concern is that the regex is a bit complex to edit, it can be constructed from a list.
```python
import re
extensions = ['txt.gz', 'xlsx']
extensions_re = re.compile(r'\.(' + '|'.join(re.escape(ext) for ext in extensions) + r')$')

# ...

out_basename = re.sub(extensions_re, '.parquet', f.name )
out_file = brick_dir / f.relative_to(download_dir).with_name( out_basename )
```

Notice that the input files each have a `v3.3` before the file suffix? Using `split('.')[0]` removes the `.3` part. This more declarative code is easier to read than a series of `split`s (`split()` assumes too much about how files are named).

Sulstice replied:
That was the concern. We can use that for later on as what the file inputs are when querying someone.

zmughal commented on stages/02_process_data.py:
```diff
@@ -0,0 +1,32 @@
+_, _, files = next(os.walk('downloads'))
+files = [ os.path.join('downloads', file) for file in files ]
+
+os.makedirs('brick', exist_ok=True)
```
```python
brick_dir = pathlib.Path('brick')
brick_dir.mkdir(exist_ok=True)
```

zmughal commented on stages/01_download_data.py:
```diff
@@ -0,0 +1,18 @@
+import os
+import urllib.request
+
+os.makedirs('downloads', exist_ok=True)
```
It's a small change, but use `download` instead of `downloads` to match up with other bricks.

zmughal commented on stages/02_process_data.py:
```diff
@@ -0,0 +1,32 @@
+              df.to_parquet(out_file)
+
+  except Exception as e:
+      print ('ADRECS File Conversion Failed: %s' % e)
```
Either get rid of the `try-except` or re`raise` the exception. We want this to indicate failure in the exit value.

Sulstice replied:
Done. should probably do a logger on this down the road.

zmughal commented on stages/01_download_data.py:
```diff
@@ -0,0 +1,18 @@
+
+os.makedirs('downloads', exist_ok=True)
+
+data_links = {
```
With a given download list, it would be better to keep the names as they are by just taking the last component of the URI path. This keeps the data closer to the original (which helps if the source ever changes) and there will be no cognitive load of looking up how this brick maps the file names.

(which is why a `wget` is a workable solution as well)

zmughal reviewed (changes requested):
Looking good. I added a couple more comments.

zmughal commented on stages/02_process_data.py:
```diff
@@ -0,0 +1,32 @@
+          reaction_data = pd.read_excel(file)
+          reaction_data.to_parquet(out_file)
+
+      elif file.endswith('gz'):
```
Perhaps be more specific about `.txt.gz`?

zmughal commented on stages/02_process_data.py:
```diff
@@ -0,0 +1,32 @@
+              data = [ line.split('\t') for line in lines ]
+              headers = data.pop(0)
+              df = pd.DataFrame(data, columns=headers)
+              df.to_parquet(out_file)
```
Add an `else` here that raises an exception to catch any unknown files.

Sulstice replied:
Done.

Sulstice committed:
changed the downloads to wget

Sulstice committed:
fixed a big in the bash script

Sulstice committed:
Added the openpyxl requirements

Sulstice committed:
Added the files update

Sulstice committed:
Added the unknown files update

Sulstice committed:
Added the extension check

zmughal commented on stages/01_download_data.sh:
```diff
@@ -0,0 +1,11 @@
+#!/usr/bin/env bash
+
+mkdir -p . /download
```
There's an extra space here between `.` and `/`

zmughal commented on stages/02_process_data.py:
```diff
@@ -0,0 +1,36 @@
+          raise Exception('Unknown File Found: %s' % file)
+
+  except:
+      print ('ADRECS File Conversion Failed: %s' % file)
```
It's easier to get rid of this outer `try-except` that way the full error can be seen. You want the error to stop the `dvc repro` process so it doesn't go any further by having a successful exit value.

zmughal committed:
Keep using `Path` instead of `str`

zmughal committed:
Replace -1 in some files with NA

since -1 means unavailable per the ADReCS data descriptions.

And treat the severity grades as categorical data.

zmughal commented:
I made a couple of changes.

The issue was with processing `ADReCS_ADR_Severity_Grade_v3.3.txt.gz`, not `Drug_ADR_v3.3.txt.gz`: that file is full of "-1" and strings such as "Mild", "Moderate" so the types were being guessed to be numeric by the reader. Adding the `low_memory=False` lets `read_csv()` see the entire file so it assigns them to strings.

I replaced both `'-1'` (str) and `-1.0` (float) with `NA` to indicate null (as written on the website) and for the severity grade, turned those columns into categorical data.

zmughal reviewed (approved):
Looks good. Just run the dvc update (replace `[s3...]`):

```
dvc repro && \
  dvc push -r [s3...]  && git add dvc.lock && git commit -m 'Update dvc.lock'
```

then merge.

Sulstice committed:
Added the dvc yaml file

Sulstice committed:
Update dvc.lock
